internal/config/config.go

```diff
@@ -0,0 +1,77 @@
+package config
+
+import (
+	"github.com/creasty/defaults"
+	"github.com/icinga/icingadb/pkg/database"
+	"github.com/icinga/icingadb/pkg/icingadb/history"
+	"github.com/icinga/icingadb/pkg/logging"
+	"github.com/icinga/icingadb/pkg/redis"
+	"github.com/pkg/errors"
+	"time"
+)
+
+// Config defines Icinga DB config.
+type Config struct {
+	Database  database.Config `yaml:"database"`
+	Redis     redis.Config    `yaml:"redis"`
+	Logging   logging.Config  `yaml:"logging"`
+	Retention RetentionConfig `yaml:"retention"`
+}
+
+func (c *Config) SetDefaults() {
+	// Since SetDefaults() is called after the default values of the struct's fields have been evaluated,
+	// setting the default port only works here because
+	// the embedded Redis config struct itself does not provide a default value.
+	if defaults.CanUpdate(c.Redis.Port) {
```

Comment:
`defaults.CanUpdate()` uses some `reflect` magic to determine this, but since that field is of type `int`, you'll never be able to distinguish whether it's set to `0` explicitly, since its zero initialised value is also `0`. So I suggest replacing this with a simple `c.Redis.Port == 0` instead. Apart from that, can you please define the default port for the database type as well, so that we no longer get [such confusing log messages](https://github.com/Icinga/icinga-go-library/issues/11).

Reply:
`internal` won't be moved to the lib so a change here won't help for that specific issue as you would have to copy `SetDefaults()` to your code. I would rather introduce something like `redis#Client.GetAddr()` and `database#DB.GetAddr()`.

Reply:
> I would rather introduce something like redis#Client.GetAddr() and database#DB.GetAddr().

That sounds good 👍!

Reply:
`redis#Client` already provides the `String()` method which exposes the addr. I also implemented `database#DB.String()` and use them for logging the initial connections:

> ...	Connecting to Database<localhost:3306 db:icingadb driver:mysql>
> ... Connecting to Redis<localhost:6380 db:0>

Reply:
IMHO that format looks strange. It seems to be chosen because `redis.Client.String()` returns `"Redis<localhost:6380 db:0>"`? Hint: You can actually get just the address, it's just that our struct definition shadows the required method, but `rc.Client.Options().Addr` should do the trick.

For a more structured logging approach, something like https://github.com/Icinga/icinga-notifications/pull/198/commits/eb18f970de441954d202f72c0641319deda27c3b could be an option.

Reply:
> defaults.CanUpdate() uses some reflect magic to determine this, but since that field is of type int, you'll never be able to distinguish whether it's set to 0 explicitly, since its zero initialised value is also 0. So I suggest replacing this with a simple c.Redis.Port == 0 instead.

No reaction to this at all?

Reply:
`defaults.CanUpdate()` and `defaults.CanUpdate()` in combination with `SetDefaults()` is the logical choice for handling default values as both is package functionality. If we ever replace this, we can look for such usages. Also, at first glance, it's pretty obvious what's happening, rather than having to recognize the initial value check. I have the impression that you are not happy with the reflection usage. But that's how the package works anyway. And if there were more such checks with different types, would you replace them with individual zero checks as well? No, because that's what this method is for.